### PR TITLE
feat: Add 3 new features to Moderator

### PR DESCRIPTION
**Pull request by JXZhou0224**
Closes # 

## 📑 Description

Three new features are add to Moderator

1. before the actual chat starts, moderator will send checking messages to all agents every 5 seconds, until all agents are fully ready to start chatting.
2. added save() method to enable moderator to save the Episode in the format of EpisodeLog
3. when chat is finished (currently, it is reaching the bound of max turns of chats), moderator will ask all agents to leave the chat

## ✅ Checks

- [x] My pull request adheres to the code style of this project
- [ ] My code requires changes to the documentation
- [ ] I have updated the documentation as required
- [ ] All the tests have passed
- [x] Branch name follows `type/descript` (e.g. `feature/add-llm-agents`)
- [x] Ready for code review

### CODE REVIEW

**Comment by JXZhou0224** [comment]
Also, for quick testing. I have added a DummyAgent in the demo file, so that the moderator can be tested without actually calling any LLM APIs.

File: sotopia/experimental/agents/moderator.py [hunk 1]
```diff
@@ -68,43 +68,143 @@ def __init__(
...
+        self.task_scheduler = asyncio.create_task(self._task_scheduler())
+        return await super().__aenter__()
+
+    async def _task_scheduler(self) -> None:
+        await self.all_agents_awake.wait()
+        return await super()._task_scheduler()
+
+    async def booting(self) -> None:
+        """
+        1. send checking message to agents for every 5 seconds, until all agents are awake
```
Comment (XuhuiZhou) [inline sotopia/experimental/agents/moderator.py:100]: can we not do like 0.1s? I think agents initiate super fast, 5s is a bit long for scaled simulations

File: sotopia/experimental/agents/moderator.py [hunk 2]
```diff
@@ -68,43 +68,143 @@ def __init__(
...
+        - if further information of the agents are needed, should be communicated through the booting process
+        2. after all agents are awake, send agent[0] a message to allow the agent to start speaking
+        """
+        while not self.all_agents_awake.is_set():
+            await self.send(
+                Observations(
+                    observations_map={
+                        output_channel: Observation(
+                            agent_name="moderator",
+                            last_turn=self.scenario,
```
Comment (XuhuiZhou) [inline sotopia/experimental/agents/moderator.py:111]: Please double check the sotopia implementation of the last turn msg here: https://github.com/sotopia-lab/sotopia/blob/61f190ee2c4b5acb94ef93c109225d6a474b4710/sotopia/envs/parallel.py#L300

The initial/boosting msg for every character should be different. This is a very important feature in Sotopia about info asymmetry

File: sotopia/experimental/agents/moderator.py [hunk 3]
```diff
@@ -68,43 +68,143 @@ def __init__(
...
+                                agent_name="moderator",
+                                last_turn=self.scenario,
+                                turn_number=0,
+                                available_actions=self.available_actions,
+                            )
+                        }
+                    )
+                )
+                break
+        self.current_agent_index += 1
```
Comment (XuhuiZhou) [inline sotopia/experimental/agents/moderator.py:142]: no need to have this here?

File: sotopia/experimental/agents/moderator.py [hunk 4]
```diff
@@ -68,43 +68,143 @@ def __init__(
...
+
+    async def save(self) -> None:
```
Comment (XuhuiZhou) [inline sotopia/experimental/agents/moderator.py:144]: maybe have a hyperparameter about whether push to DB? and return epilog instead?

**Review by XuhuiZhou — CHANGES_REQUESTED** [review]
Thanks for the effort! left some comments. One structural thing to note is that let's not use moderator inherit from the baseAgent.

Since the interface of baseAgent should be different. And we would want moderator have different event loop than the agent in the future.

This could be easy to do for now, just copy the baseagent code but make sure the send does not imply the action has to be the same

**Comment by XuhuiZhou** [comment]
> Also, for quick testing. I have added a DummyAgent in the demo file, so that the moderator can be tested without actually calling any LLM APIs.

For testing purpose, can we route to the pytest files?

File: sotopia/experimental/agents/moderator.py (hunk 3 above)
Comment (JXZhou0224) [reply sotopia/experimental/agents/moderator.py:144]: Yes, I think it is needed under the current implementation, as the booting function will give the first agent permission to speak. Therefore, on receiving the next message, moderator should be asking the next agent to speak

File: sotopia/experimental/agents/moderator.py (hunk 2 above)
Comment (JXZhou0224) [reply sotopia/experimental/agents/moderator.py:113]: I have made a temprary fix for this. I also notice that the current ScriptBackGround dataType only supports 2 agents in chat. I think I will create a new background dataType for moderator.

**Comment by JXZhou0224** [comment]
Thanks for the review! I have made a few changes:
1. now the moderator no longer inherits from BaseAgent but from Node directly. Passing the mypy checks.
2. it is now possible to set different background for different agents

File: sotopia/experimental/agents/moderator.py (hunk 2 above)
Comment (XuhuiZhou) [reply sotopia/experimental/agents/moderator.py:141]: thanks!

**Review by XuhuiZhou — APPROVED** [review]

**Pull request merged**